include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -179,6 +179,30 @@ __pattern_copy(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _InRa
         oneapi::dpl::__ranges::views::all_read(std::forward<_InRange>(__in_r)),
         oneapi::dpl::__ranges::views::all_write(std::forward<_OutRange>(__out_r)));
 }
+
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R, typename _T>
+std::ranges::borrowed_iterator_t<_R>
+__pattern_fill(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r, const _T& __value)
+{
+    auto __f = [__value](auto&& __a) { __a = __value; };
+
+    oneapi::dpl::__internal::__ranges::__pattern_walk_n(__tag, std::forward<_ExecutionPolicy>(__exec), __f,
+                                                            oneapi::dpl::__ranges::views::all(std::forward<_R>(__r)));
+
+    return {std::ranges::begin(__r) + std::ranges::size(__r)};
```

Comment:
If we cant rely upon `std::ranges::end(r)` to be the type we want, does it make sense to make a helper like `oneapi::dpl::ranges::__internal::end_iterator(Range&& r)` which returns `begin(r)  + size(r)`? 

 We do this in a lot of places.

Reply:
I duno..
On one hand - it makes sense.. On the other hand, "begin+ size" so trivial expression and has good readability..
In case of "end_iterator" a user should find a "end_iterator" function definition to understand the logic.
And finally, an element of irony  - lengths of the lines are equal)

`{std::ranges::begin(__r) + std::ranges::size(__r)}`
`oneapi::dpl::ranges::__internal::end_iterator(__r)`

Reply:
Basically, returning the last iterator (not sentinel) for range based serial implementation - is the native way and the last is calculated within serial algorithm "natively".
For parallel implementation - "it depends".... so we have an universal approach for (RA ranges) - `{std::ranges::begin(__r) + std::ranges::size(__r)}`.

Reply:
Yeah, I'm fine with it as is.  Resolving.